### PR TITLE
when discovery selector changed, it should sync namespace ca configmap

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Fix https://github.com/istio/istio/issues/44675

It need further refactor

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
@howardjohn when working on this, i found ingress/gateway/other istio configs, we donot need to trigger cb to handle them directly. They are now reconciled indirectly when istiod doing full push

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 1]
```diff
@@ -69,8 +72,8 @@ func NewNamespaceController(kubeClient kube.Client, caBundleWatcher *keycertbund
 		controllers.WithReconciler(c.insertDataForNamespace),
 		controllers.WithMaxAttempts(maxRetries))
 
-	c.configmaps = kclient.New[*v1.ConfigMap](kubeClient)
-	c.namespaces = kclient.New[*v1.Namespace](kubeClient)
+	c.configmaps = kclient.NewFiltered[*v1.ConfigMap](kubeClient, kclient.Filter{ObjectFilter: c.GetFilter()})
+	c.namespaces = kclient.NewFiltered[*v1.Namespace](kubeClient, kclient.Filter{ObjectFilter: c.GetFilter()})
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:76]: This change is the cause of test failure

File: pilot/pkg/serviceregistry/kube/controller/controller.go [hunk 2]
```diff
@@ -267,10 +267,8 @@ func NewController(kubeClient kubelib.Client, options Options) *Controller {
 		)
 	}
 
-	if c.opts.DiscoveryNamespacesFilter == nil {
-		c.opts.DiscoveryNamespacesFilter = namespace.NewDiscoveryNamespacesFilter(c.namespaces, options.MeshWatcher.Mesh().DiscoverySelectors)
-	}
-
+	// always init for each cluster, otherwise different ns labels in different cluster may not take effect
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:270]: Yes, i try to expalain a little bit. The first time init in multicluster only watches the config cluster's namespaces.  In real scenario, the remote clusters may have different namespace with different labels. 
I think it is a bug if the namespace sameness across clusters can not be guranteed.

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 3]
```diff
@@ -162,9 +172,24 @@ func (nc *NamespaceController) syncNamespace(ns *v1.Namespace) {
...
+// handleNamespace called on namespace selected/deselected by selector change
+func (nc *NamespaceController) handleNamespace(ns string, event model.Event) {
+	// skip special kubernetes system namespaces
+	if inject.IgnoredNamespaces.Contains(ns) {
 		return
 	}
-	nc.queue.Add(types.NamespacedName{Name: ns.Name})
+	if event == model.EventDelete {
+		// delete the configmap
+		if err := nc.configmaps.Delete(CACertNamespaceConfigMap, ns); err != nil {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:186]: make sense, i found it need some refactor to enqueue

**Comment by howardjohn** [comment]
One note - this does not fix https://github.com/istio/istio/issues/44675, only solves part of it. There is still a lot of other problems with the filter.

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:270]: So with this PR, I ahve just a single cluster but I get 2 DiscoveryNamespacesFilter created.

```
2023-05-10T22:15:52.120764Z     error   howardjohn: building discovery filter goroutine 1 [running]:
runtime/debug.Stack()
        /home/howardjohn/sdk/go1.20/src/runtime/debug/stack.go:24 +0x65
istio.io/istio/pkg/kube/namespace.NewDiscoveryNamespacesFilter({0x37319a0?, 0xc000877100}, {0x0, 0x0, 0x0})
        /home/howardjohn/go/src/istio.io/istio/pkg/kube/namespace/filter.go:69 +0x47
istio.io/istio/pkg/kube/multicluster.NewController({0x373eee8, 0xc0008a0420}, {0x3233bcd, 0xc}, {0x322e580, 0xa}, {0x3711fa0, 0xc000b7d3b0})
        /home/howardjohn/go/src/istio.io/istio/pkg/kube/multicluster/secretcontroller.go:157 +0x709
istio.io/istio/pilot/pkg/bootstrap.(*Server).initMulticluster(0xc000908e00, 0x2ad2d20?)
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:1136 +0x5d
istio.io/istio/pilot/pkg/bootstrap.(*Server).initControllers(0xc000908e00, 0xc000ad5680)
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:1115 +0x4e
istio.io/istio/pilot/pkg/bootstrap.NewServer(0xc000ad5680, {0x0, 0x0, 0x0?})
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:298 +0xcec
istio.io/istio/pilot/cmd/pilot-discovery/app.newDiscoveryCommand.func2(0xc000b02000?, {0x54e7f38?, 0x0?, 0x0?})
        /home/howardjohn/go/src/istio.io/istio/pilot/cmd/pilot-discovery/app/cmd.go:100 +0x4e
github.com/spf13/cobra.(*Command).execute(0xc000b02000, {0x54e7f38, 0x0, 0x0})
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc000a87b00)
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(0x0?)
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992 +0x19
main.main()
        /home/howardjohn/go/src/istio.io/istio/pilot/cmd/pilot-discovery/main.go:27 +0x25

```

And

```
2023-05-10T22:15:52.244468Z     error   howardjohn: building discovery filter goroutine 1 [running]:
runtime/debug.Stack()
        /home/howardjohn/sdk/go1.20/src/runtime/debug/stack.go:24 +0x65
istio.io/istio/pkg/kube/namespace.NewDiscoveryNamespacesFilter({0x37319a0?, 0xc0009e5040}, {0x0, 0x0, 0x0})
        /home/howardjohn/go/src/istio.io/istio/pkg/kube/namespace/filter.go:69 +0x47
istio.io/istio/pilot/pkg/serviceregistry/kube/controller.NewController({0x373eee8?, 0xc0008a0420}, {{0x3233bcd, 0xc}, 0xc000aff760, {0x3235ae9, 0xd}, {0x322e580, 0xa}, 0xc000a175f0, ...})
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/serviceregistry/kube/controller/controller.go:271 +0x73e
istio.io/istio/pilot/pkg/serviceregistry/kube/controller.(*Multicluster).addCluster(0xc000a01380, 0xc000b80420)
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/serviceregistry/kube/controller/multicluster.go:219 +0x268
istio.io/istio/pilot/pkg/serviceregistry/kube/controller.(*Multicluster).ClusterAdded(0xc000a01380, 0x0?, 0xc0005cbad8?)
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/serviceregistry/kube/controller/multicluster.go:157 +0x88
istio.io/istio/pkg/kube/multicluster.(*Controller).handleAdd(0x0?, 0xc000a90018?, 0x2ad22e0?)
        /home/howardjohn/go/src/istio.io/istio/pkg/kube/multicluster/secretcontroller.go:424 +0x57
istio.io/istio/pkg/kube/multicluster.(*Controller).Run(0xc0008926c0, 0xc000087740)
        /home/howardjohn/go/src/istio.io/istio/pkg/kube/multicluster/secretcontroller.go:175 +0x89
istio.io/istio/pilot/pkg/bootstrap.(*Server).initMulticluster.func1(0xc000112120?)
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:1139 +0x27
istio.io/istio/pilot/pkg/server.(*instance).Start(0xc000a4ed00, 0xc000087740)
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/server/instance.go:82 +0x11d
istio.io/istio/pilot/pkg/bootstrap.(*Server).Start(0xc000908e00, 0x0?)
        /home/howardjohn/go/src/istio.io/istio/pilot/pkg/bootstrap/server.go:432 +0xd4
istio.io/istio/pilot/cmd/pilot-discovery/app.newDiscoveryCommand.func2(0xc000b02000?, {0x54e7f38?, 0x0?, 0x0?})
        /home/howardjohn/go/src/istio.io/istio/pilot/cmd/pilot-discovery/app/cmd.go:106 +0xa5
github.com/spf13/cobra.(*Command).execute(0xc000b02000, {0x54e7f38, 0x0, 0x0})
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc000a87b00)
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(0x0?)
        /home/howardjohn/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992 +0x19
main.main()
        /home/howardjohn/go/src/istio.io/istio/pilot/cmd/pilot-discovery/main.go:27 +0x25
```

(added some logging to get this)

This seems unexpected? I should only have 1 with single cluster

**Review by howardjohn — COMMENTED** [review]
Generally looks good, mostly just concerned about the 2x creation.

I think long term we should fix this more holistically in `kclient`, as described in https://github.com/istio/istio/issues/44675. Each controller shouldn't have to worry about this. But for now I think this is an ok first step

File: pilot/pkg/serviceregistry/kube/controller/controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:270]: Yes, i think need to skip it for the config cluster

File: pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go [hunk 4]
```diff
@@ -132,15 +145,26 @@ func (nc *NamespaceController) startCaBundleWatcher(stop <-chan struct{}) {
...
-func (nc *NamespaceController) insertDataForNamespace(o types.NamespacedName) error {
+// If the namespace is filtered out by discovery selector, the configmap will be deleted.
+func (nc *NamespaceController) reconcileCACert(o types.NamespacedName) error {
 	ns := o.Namespace
 	if ns == "" {
 		// For Namespace object, it will not have o.Namespace field set
 		ns = o.Name
 	}
+
+	if nc.DiscoveryNamespacesFilter != nil && !nc.DiscoveryNamespacesFilter.GetMembers().Contains(ns) {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go:158]: Yes, will update

File: pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go [hunk 5]
```diff
@@ -36,53 +34,35 @@ func (c *Controller) initDiscoveryHandlers(meshWatcher mesh.Watcher, discoveryNa
...
-							Reason: []model.TriggerReason{model.NamespaceUpdate},
-						})
-					}
-					return nil
+	discoveryNamespacesFilter.AddHandler(func(ns string, event model.Event) {
+		switch event {
+		case model.EventAdd:
+			c.queue.Push(func() error {
+				c.handleSelectedNamespace(ns)
+				// This is necessary because namespace handled by discoveryNamespacesFilter may take some time,
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go:42]: Do we really need this? I think we already do a full push in `SelectorsChanged` ?

File: pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go [hunk 6]
```diff
@@ -171,6 +151,6 @@ func (c *Controller) handleDeselectedNamespace(ns string) {
 	}
 
 	if err := multierror.Flatten(errs.ErrorOrNil()); err != nil {
-		log.Errorf("one or more errors while handling delabeled discovery namespace %s: %v", ns, err)
+		log.Errorf("one or more errors while handling deselected discovery namespace %s: %v", ns, err)
```
Comment (ramaraochavali) [inline pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go:154]: nit: change same comment in L 138

File: pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go:42]: Let me take a look though it is from previous patch

File: pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go (hunk 6 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go:154]: ok

File: pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go:42]: please look at the latest commit

[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go:42]:
```
meshWatcher.AddMeshHandler(func() {
		discoveryNamespacesFilter.SelectorsChanged(meshWatcher.Mesh().GetDiscoverySelectors())
}
```

Here it only calls selectorchanges, anywork else is done in the implement, which run the handlers we registred here

[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go:42]: The handler is also triggered when a namespace object updates

[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/discoverycontrollers.go:42]: I think it makes sense and looks lot better.

**Comment by costinm** [comment]
I am a bit worried about how much we watch and do for multi-cluster.

Clusters are supposed to be admin separation boundaries, stronger than
namespace boundaries.  "Sameness" for identity and naming - and
having (some or all) services load balance across all clusters is already
tricky from a security/blast radius and behavior perspective. Maybe it
is too late to change for normal Istio - but for Ambient and Safe Istio we
should have a review of the multi-cluster design and security,
and at least document the security implications and what users can do to
reduce them.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**